[Deepspeed] adapt multiple models, add zero_to_fp32 tests (#12477)
huggingface/transformers: docs/source/main_classes/deepspeed.rst

While the fp16 weights are fine for resuming training, if you finished finetuning your model and want to upload it to
the `models hub <https://huggingface.co/models>`__ or pass it to someone else you most likely will want to get the fp32
weights. This ideally shouldn't be done during training since this is a process that requires a lot of memory, and
therefore best to be performed offline after the training is complete. But if desired and you have plenty of free CPU
memory it can be done in the same training script. The following sections will discuss both approaches.

**Live FP32 Weights Recovery:**

This approach may not work if you model is large and you have little free CPU memory left, at the end of the training.

If you have saved at least one checkpoint, and you want to use the latest one, you can do the following:

.. code-block:: python

    from transformers.trainer_utils import get_last_checkpoint
    from deepspeed.utils.zero_to_fp32 import load_state_dict_from_zero_checkpoint
    checkpoint_dir = get_last_checkpoint(trainer.args.output_dir)
    fp32_model = load_state_dict_from_zero_checkpoint(trainer.model, checkpoint_dir)

> If you're using the ``--load_best_model_at_end`` class:`~transformers.TrainingArguments` argument (which disabled
> checkpoint saving), then you can finish the training by first saving a checkpoint explicitly and then do the same as

Reviewer: The `--load_best_model_at_end` does not disable checkpoint saving at all, it's just that in that case, the best checkpoint is not necessary the best one.
```suggestion
If you're using the ``--load_best_model_at_end`` class:`~transformers.TrainingArguments` argument (to track the best checkpoint), then you can finish the training by first saving the final model explicitly and then do the same as
```

Author: That wasn't my experience. I was getting no checkpoint saved with `--load_best_model_at_end` - I will need to re-tests that this is indeed so. I will re-check and revert to you.

Author: I double checked - `--load_best_model_at_end` disables checkpoint saving.

Author: This saves:

```
export BS=16; rm -r output_dir; PYTHONPATH=src USE_TF=0 CUDA_VISIBLE_DEVICES=0,1 python -m torch.distributed.launch --nproc_per_node=2 examples/pytorch/translation/run_translation.py --model_name_or_path t5-small --output_dir output_dir --adam_eps 1e-06  --do_train --label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step --logging_steps 500 --max_source_length 128 --max_target_length 128 --val_max_target_length 128 --num_train_epochs 1 --overwrite_output_dir --per_device_train_batch_size $BS --predict_with_generate --sortish_sampler --source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" --source_prefix "translate English to Romanian: " --warmup_steps 50 --max_train_samples 50 --save_steps 1 
```

then adding `--load_best_model_at_end`

no longer saves the checkpoints.

Author: You wrote:
> The --load_best_model_at_end does not disable checkpoint saving at all, it's just that in that case, the best checkpoint is not necessary the best one.

Is there a typo there? 
"the best checkpoint is not necessary the best one." 
Did you mean to say: 
"the best checkpoint is not necessary the last one"
?

Reviewer: `--load_best_model_at_end` requires evaluations to happen (otherwise it does not know which model is the best) and overrides the saving strategy to be the same as the evaluation strategy. Your command does not comport evaluation (no `--do_eval`) so that's why it does not save anything.

> Is there a typo there?
> "the best checkpoint is not necessary the best one."
> Did you mean to say:
> "the best checkpoint is not necessary the last one"
> ?

Yes, that's what I meant.

Author: Thank you for clarifying.

Won't the simplest solution here be to assert and require `do_eval` if `--load_best_model_at_end` is passed w/o the former? 

Silently overriding `--save_steps` is a difficult one. 

A somewhat similar ambiguous action-at-a-distance is `evaluation_strategy` activates eval... (not the same but also a surprise). Often, there are many args and one would think that removing `do_eval` should disable eval, but it's not, so it's not easy on the user side.

Author: But it doesn't change your proposed comment, it's still a great suggestion - I will commit it once this conversation is over, otherwise github will make it disappear.

Author: @sgugger, when you get a chance let's conclude this one - or perhaps might be easier to start a new issue as it's unrelated to this PR.

Reviewer: We can look at that in another PR, this is not directly linked to this one. I don't think we should entangle  `--do_eval` which means evaluate my model after training inside the example scripts with the training arguments used by the `Trainer`, but we can definitely have `--load_best_model_at_end` assert on `eval_strategy` behind different from NO.

Author: https://github.com/huggingface/transformers/issues/12685

above:

.. code-block:: python

    from deepspeed.utils.zero_to_fp32 import load_state_dict_from_zero_checkpoint
    checkpoint_dir = os.path.join(trainer.args.output_dir, "checkpoint-final")
    trainer.deepspeed.save_checkpoint(checkpoint_dir)
    fp32_model = load_state_dict_from_zero_checkpoint(trainer.model, checkpoint_dir)

.. note::

    Note, that once ``load_state_dict_from_zero_checkpoint`` was run, the ``model`` will no longer be useable in the
    deepspeed context of the same application. i.e. you will need to re-initialize the deepspeed engine, since
    ``model.load_state_dict(state_dict)`` will remove all the deepspeed magic from it. So do this only at the very end

>     of the training.

Reviewer: deepspeed -> DeepSpeed ?
